### PR TITLE
Improving context stack

### PR DESCRIPTION
**Pull request by sklam**
based on https://github.com/numba/numba/pull/6762

Improves the context stack implementation in #6762 and refactor the target options and flags.

### CODE REVIEW

**Comment by sklam** [comment]
This will need CUDA smoketest when done

**Comment by stuartarchibald** [comment]
https://github.com/stuartarchibald/numba/commit/4666c774721cc8225f2c7a6dce096a6b68cd1caa hacks through the flags for `operator.add`.

**Comment by sklam** [comment]
There're known cuda error due to changes to error msg. To be fixed later.

**Comment by sklam** [comment]
@stuartarchibald, 

> stuartarchibald@4666c77 hacks through the flags for operator.add.

looks promising. I added a print of the function name and the flags in the lowerer and it showing the right things:

```
----- jit_wrapper__built_in_function_add_ Flags(enable_looplift=True, enable_pyobject=False, enable_pyobject_looplift=True, debuginfo=False, boundscheck=False, no_cpython_wrapper=True, nrt=True, fastmath=FastMathOptions({'fast'}))
----- foo Flags(enable_looplift=True, enable_pyobject=False, enable_pyobject_looplift=True, debuginfo=False, boundscheck=False, nrt=True, fastmath=FastMathOptions({'fast'}))
```

for 

```python
from numba import njit

@njit(fastmath=True)
def foo(x, y):
  a = x + y
  return a


x, y = 1, 2
r = foo(x, y)
print(r)
```

File: numba/core/targetconfig.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            def delter(self):
+                del self._values[name]
+
+            return property(getter, setter, delter, option.doc)
+
+        for name, option in cls.options.items():
+            setattr(cls, name, make_prop(name, option))
+
+    def find_options(cls, dct):
+        """Returns a new dict with all the items that is mapping to a
```
Comment (stuartarchibald) [inline numba/core/targetconfig.py:53]:
```suggestion
        """Returns a new dict with all the items that are a mapping to an
```

File: numba/core/targetconfig.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+class TargetConfig(metaclass=_MetaTargetConfig):
+    def __init__(self, copy_from=None):
+        self._values = {}
+        if copy_from is not None:
+            assert isinstance(copy_from, TargetConfig)
+            self._values.update(copy_from._values)
+
+    def __repr__(self):
+        # NOTE: defaulted options will be placed in the back and grouped inside
+        #       a square bracket; i.e. [optname=optval, ...]
```
Comment (stuartarchibald) [inline numba/core/targetconfig.py:68]:
```suggestion
        # NOTE: default options will be placed at the end and grouped inside
        #       a square bracket; i.e. [optname=optval, ...]
```

File: numba/core/utils.py [hunk 3]
```diff
@@ -177,6 +180,38 @@ def shutting_down(globals=globals):
 atexit.register(_at_shutdown)
 
 
+class ConfigStack:
```
Comment (stuartarchibald) [inline numba/core/utils.py:183]: Docstr?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @sklam, looks good, just a few minor things to resolve.

**Comment by stuartarchibald** [comment]
I've also checked that this does transmit the flags correctly, applying this patch: https://github.com/stuartarchibald/numba/commit/2f63896bd848971c60393aece5f1371a46c5f637 will get the flags to the right place and also disables the `fastmath` visitor. Then running something like:
```python
from numba import njit, types
import numpy as np
import operator

def foo(n):
    return operator.add(n, 3.14)

# no fast
jfoo = njit(foo)
jfoo.compile((types.float64,),)
print(jfoo.inspect_llvm(jfoo.signatures[0]))
```
with and without `fastmath=True` in the `jit` decorator demonstrates the presence (or not) of a `fadd fast` in the LLVM IR.

File: numba/core/base.py [hunk 4]
```diff
@@ -834,10 +834,19 @@ def _compile_subroutine_no_cache(self, builder, impl, sig, locals={},
...
+                from numba.core import utils
+
+                cstk = utils.ConfigStack()
                 flags = compiler.Flags()
-            flags.set('no_compile')
-            flags.set('no_cpython_wrapper')
-            flags.set('no_cfunc_wrapper')
+                if cstk:
+                    tls_flags = cstk.top()
+                    if tls_flags.is_set("nrt") and tls_flags.nrt:
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/base.py:843]: `tls_flags.nrt` by itself can be reading the default. Here, it is checking that "nrt" is not default and its value is `True`.

File: numba/core/ccallback.py [hunk 5]
```diff
@@ -16,14 +16,14 @@
 class _CFuncCompiler(_FunctionCompiler):
 
     def _customize_flags(self, flags):
-        flags.set('no_cpython_wrapper', True)
-        flags.set('no_cfunc_wrapper', False)
+        flags.no_cpython_wrapper = True
+        flags.no_cfunc_wrapper = False
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/ccallback.py:20]: agree

File: numba/core/base.py [hunk 6]
```diff
@@ -834,10 +834,19 @@ def _compile_subroutine_no_cache(self, builder, impl, sig, locals={},
             codegen = self.codegen()
             library = codegen.create_library(impl.__name__)
             if flags is None:
+                from numba.core import utils
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/base.py]: Removed import. It's already imported at the top

File: numba/core/targetconfig.py [hunk 7]
```diff
@@ -105,13 +157,24 @@ def is_set(self, name):
...
     def copy(self):
+        """Clone this instance.
+        """
         return type(self)(self)
 
     def summary(self):
+        """Returns a ``str`` that summarizes this instance.
+
+        In contrast to ``__repr__``, only options that are explicitly set will
+        be showed.
```
Comment (stuartarchibald) [inline numba/core/targetconfig.py:176]:
```suggestion
        be shown.
```

**Review by stuartarchibald — DISMISSED** [review]
Thanks for re-implementing and cleaning this up @sklam, the context stack is really useful and the side effect of improving the flag system is great!

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_34`.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cpu_yaml_13`.

**Comment by stuartarchibald** [comment]
Both CPU and CUDA tests are running through the farm.

**Comment by stuartarchibald** [comment]
CUDA tests are failing on all platforms with:
```
ERROR: test_target_cuda_unrecognized_arg (numba.cuda.tests.cudapy.test_vectorize_decor.TestVectorizeUnrecognizedArg)
<traceback>
KeyError: "cuda vectorize target does not support option: 'nonexistent'"
```

**Comment by sklam** [comment]
started new cuda smoketest: `numba_smoketest_cuda_yaml_35`

File: numba/core/options.py [hunk 8]
```diff
@@ -8,27 +8,41 @@
...
+    """
     class Mapping:
         def __init__(self, flag_name, apply=lambda x: x):
             self.flag_name = flag_name
             self.apply = apply
 
-    def finalize(self, settings, flags):
+    def finalize(self, flags, options):
+        """Subclass can override this method to make target specific
+        customization of default flags.
```
Comment (stuartarchibald) [inline numba/core/options.py:21]:
```suggestion
        """Subclasses can override this method to make target specific
        customizations of default flags.
```

File: numba/tests/npyufunc/test_vectorize_decor.py [hunk 9]
```diff
@@ -129,12 +129,12 @@ def _test_target_unrecognized_arg(self, target, with_sig=True):
...
         sig = [float32(float32, float32)]
         args = with_sig and [sig] or []
-        with self.assertRaises(NameError) as raises:
+        with self.assertRaises(KeyError) as raises:
             f = vectorize(*args, target=target, nonexistent=2)(vector_add)
             f(a, b)
         self.assertIn("Unrecognized options", str(raises.exception))
 
-class TestVectorizeUnrecognizedArg(BaseVectorizeUnrecognizedArg):
+class _(BaseVectorizeUnrecognizedArg):
```
Comment (stuartarchibald) [inline numba/tests/npyufunc/test_vectorize_decor.py:137]: Is this name change intended?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the fixes, change set from ec69a18 onwards is checked, few minor things to resolve else looks good.

File: numba/core/options.py [hunk 10]
```diff
@@ -1,97 +1,103 @@
...
-            try:
-                ctor = self.OPTIONS[k]
-            except KeyError:
-                fmt = "%r does not support option: '%s'"
-                raise KeyError(fmt % (self.__class__, k))
-            else:
-                self.values[k] = ctor(v)
+    def finalize(self, flags, options):
+        """Subclass can override this method to make target specific
+        customization of default flags.
```
Comment (stuartarchibald) [inline numba/core/options.py:21]:
```suggestion
        """Subclasses can override this method to make target specific
        customizations of default flags.
```

File: numba/tests/npyufunc/test_vectorize_decor.py (hunk 9 above)
Comment (sklam) [reply numba/tests/npyufunc/test_vectorize_decor.py:137]: oh no

Comment (sklam) [reply numba/tests/npyufunc/test_vectorize_decor.py]: fixed now https://github.com/numba/numba/pull/6814/commits/e5497c1fb0b61996c11de55f18ce72963ebefb58

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fixes!

File: numba/tests/npyufunc/test_vectorize_decor.py (hunk 9 above)
Comment (stuartarchibald) [reply numba/tests/npyufunc/test_vectorize_decor.py]: thanks

**Comment by stuartarchibald** [comment]
> started new cuda smoketest: `numba_smoketest_cuda_yaml_35`

This passed on the 5/6 of the CUDA test matrix that ran (the rest suffered an intermittent unrelated problem) and given the nature of the patch/previous failures, that any part of the matrix passed is sufficient confirmation of the fixes. Thanks again for all your work on this, it should unlock a load more optimisation opportunities :) Marking ready to merge...!

**Comment by sklam** [comment]
Azure CI build https://github.com/numba/numba/pull/6814/commits/24b16a6234ed69f6a9f49bf909d6f5c7181a965b twice. The first time is all green https://dev.azure.com/numba/numba/_build/results?buildId=8387&view=results. The second time it got stuck in internet traffic jam.

**Pull request merged**